src/main/java/org/betonquest/betonquest/api/PlayerObjectiveEndEvent.java

```diff
@@ -0,0 +1,107 @@
+package org.betonquest.betonquest.api;
+
+import org.betonquest.betonquest.id.ObjectiveID;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+
+/**
+ * When player end event.
+ */
+public class PlayerObjectiveEndEvent extends PlayerEvent {
+
+    /**
+     * HandlerList of this event.
+     */
+    private static final HandlerList HANDLERS = new HandlerList();
+
+    /**
+     * Objective which is just ended of this event.
+     */
+    private final Objective objective;
+
+    /**
+     * Reason of why objective end.
+     */
+    private final EndCause endCause;
+
+    /**
+     * Constructor of PlayerObjectiveEndEvent.
+     *
+     * @param who player who end the event
+     * @param objective the objective which just ended
+     * @param endCause why is the objective end
+     */
+    public PlayerObjectiveEndEvent(final Player who, final Objective objective, final EndCause endCause) {
+        super(who);
+        this.objective = objective;
+        this.endCause = endCause;
+    }
+
+    /**
+     * Reason enum of why objective end.
+     */
+    public enum EndCause {
+
+        /**
+         * Objective end because it is completed.
+         */
+        COMPLETE,
+
+        /**
+         * Objective end because it is canceled.
+         */
+        CANCEL,
+
+        /**
+         * Objective end because it is paused.
+         */
+        PAUSE
+    }
```

Comment:
I don't think that this solution with an enum that describes the changes is good. This enum couples the objective change events to so many other classes.

How about an enum (possibly as inner class of `Objective`) that represents the current state:
```java
enum ObjectiveState {
    NEW,
    ACTIVE,
    COMPLETED,
    PAUSED,
    CANCELED,
}
```
Or implement the [state pattern](https://en.wikipedia.org/wiki/State_pattern)? I don't know if it fits here, maybe I'll do it myself later.

The objective change events could then just provide the previous an new state. You could even just make one `ObjectiveChangeEvent` instead of the current two then.

Of course I am open for different suggestions if you happen to have a better idea.

Reply:
Thanks for your point, but I think this solution has some problems.

First of all, there is only one Objective instance for one Objective, and all players share this instance, but there is an ObjectiveData class. Each Objective class of each player should have an ObjectiveData instance. If I want to add an ObjectiveState enum, Then this state enum should belong to ObjectiveData instead of ObjectiveState.

Secondly, the ObjectiveData instance only exists when a player’s Objective is in the ACTIVE state. Once it enters the COMPLETED, PAUSED or CANCELED state, the ObjectiveData object will be removed from objective.dataMap and then no longer exists. As for the NEW state, it should It only exists for a short time after the object is initialized and before the event is thrown, and then it is replaced by the ACTIVE state. So if we don't get the new state of ObjectiveData in the event, no matter we take out an ObjectiveData from objective.dataMap and .getState() at any time, we can only get an ACTIVE state, and all other states will only be in the event.

In addition to objective.dataMap, there is also player objective data in the database. We can indeed put a Map<Player, ObjectiveState> in the Objective class, but currently only ACTIVE and PAUSED Objectives will be saved in the database. so if no code changes are made, when the server restarts, there will only be two states of ACTIVE and PAUSED in this Map.

Reply:
I think, if we just want to show the reasons of the start and end of Objective for the event, then it is also possible to add StartCause and EndCause to the event.

Reply:
I see your point with adding the enum. I didn't mean to set the state inside the ObjectiveDate (or even the Objective). What I meant was that instead of providing an enum value representing a change, two enum values that represent the previous and future state could be used.

A method like [`resumeObjective()`](https://github.com/BetonQuest/BetonQuest/pull/1536/files#diff-c9453a4f3323997752f74c1c4fbbd21a5f77ec9504721230aebf0cf3740a06deR443-R466) should either know about the event and fire it itself or not know about it at all, but never just know about the `PlayerObjectiveStartEvent.StartCause` enum.

> I can understand your concern, I added `removePlayer(final String playerID, final PlayerObjectiveEndEvent.EndCause endCause)` and `addPlayer(final String playerID, final String instructionString, final PlayerObjectiveStartEvent.StartCause startCause)` methods […], the use of new methods in other classes is just to make the code look more accurate.

These two methods are exactly my concern. There is no reason for any client starting an objective to know about the event.

> […] but I also keep `removePlayer(final String playerID)` and `addPlayer (final String playerID, final String instructionString)` these two old methods to keep api […].

You are contributing to BetonQuest 2.0 dev, so API breaking changes are allowed, in fact there are major API changes planned. Though we're currently mostly occupied by the comment support that @Wolf2323 contributes to snakeyml and  Bukkit (and thus Spigot, Paper and friends).

Reply:
I understand what you mean now. I agree with that. I commited the latest code. Please check it out.